server/src/main/java/org/apache/druid/curator/CuratorModule.java

```diff
@@ -79,10 +80,24 @@ public CuratorFramework makeCurator(CuratorConfig config, EnsembleProvider ensem
           StringUtils.format("%s:%s", config.getZkUser(), config.getZkPwd()).getBytes(StandardCharsets.UTF_8)
       );
     }
+
+    final Function<Void, Void> exitFunction = new Function<Void, Void>()
+    {
+      @Override
+      public Void apply(Void someVoid)
+      {
+        log.error("Zookeeper can't be reached, forcefully stopping lifecycle...");
+        lifecycle.stop();
```

Comment:
Are you sure that this function will actually run before the function configured in `makeEnsembleProvider()` will terminate the process?

Reply:
I'm not sure I understand the question.  Can you please rephrase?

Reply:
`EnsembleProvider` created in `makeEnsembleProvider()` method, where you also added exit (but without stopping a `lifecycle`), is injected only in this method. So they are somehow related. I don't know what `EnsembleProvider` provider is and which termination action will kick in first.

Reply:
When I was testing this, I could never make the makeEnsembleProvider(...)-created RetryPolicy actually fire.  It was always the makeCurator(...)-created RetryPolicy that fired (and exited).  I'm just covering all the bases.

Reply:
It is clearly logged:

makeCurator()'s exit will look like this:

> Zookeeper can't be reached, forcefully stopping lifecycle...
> Zookeeper can't be reached, forcefully stopping virtual machine...


makeEnsembleProvider()'s exit will look like this:

> Zookeeper can't be reached, forcefully stopping virtual machine...